### PR TITLE
Pull dedupe out of queue and into personality mixin

### LINKED ISSUE #331: [Deduplication] look at extracting deduplication logic from sequencer queue into mixin
We still have some logic in the internal queue used for sequencing that performs some deduplication: https://github.com/transparency-dev/trillian-tessera/blob/main/storage/internal/queue.go#L102

This is a bit of a funny situation as it's very internal to Tessera, and not easily controllable by users. It also has some peculiarities because it only dedupes within an in-memory batch. This means the following code could lead to 1 or 2 entries being in the log, depending on whether a batch was flushed in between the calls:

```
s.Add(x)
s.Add(x)
```

Take a look at pulling out this deduplication into a mixin similar to the `Awaiter`. This will let personalities have more control over how this works, and Tessera will have less code, and be more predictable.

### PR DESCRIPTION
**Pull request by mhutchinson**
This allows the deduplication behaviour to be more precisely controlled by the personality. Previously deduping was always performed within a single batch flushed from the queue, and this could not be disabled. Now, deduping can be disabled, or made to apply across batches if a large enough cache size is set.

The conformance binaries have been updated to use this new mixin, as this is the closest to a no-op change I can make. We may want to look at making the cache size configurable and see how this affects performance.

This fixes #331.

### CODE REVIEW

**Comment by phbnf** [comment]
There's already some deduplication code that was written for CT, it's Best Effort deduplication that eventually converges on always returning the smallest index possible for a submitted leaf. The plan is to take what we can out of the static-ct repo later on, and put it back in this tessera repo. This PR will change how to reason about this, both for in terms of how we align  the existing and new deduplication APIs, and what happens if both deduplications are chained together (I think it will make convergence properties slightly different). Sounds good to me to move forward with it if you want this out before Alpha, and we can re-think how the personality mixins work with one another later?

**Comment by mhutchinson** [comment]
> Sounds good to me to move forward with it if you want this out before Alpha, and we can re-think how the personality mixins work with one another later?

Great I'll merge this if I can get an approval. Until we hit 1.0 all parts of the API are still open to change. Ideally the core parts will change less, and the mixins are more open to change. So sounds like we're aligned that this mixin works for now but can be changed if needed.

FWIW I really like the approach I've taken here of making the mixins that modify the `Add` call have the same signature so you can use them as a functional decorator pattern. But if it needs to change then it can. 🙃

File: cmd/conformance/aws/main.go
Comment (roger2hk) [inline cmd/conformance/aws/main.go:85]: I find it a bit weird when calling the `dedupe.Add`. It sounds like it's adding the entry to the dedupe queue instead of the storage. Can we rename `dedupe` to something else like `dedupedStorage`?

**Review by roger2hk — APPROVED** [review]
I was expecting the dedupe mixin to be defined as one of the `*options.StorageOptions`. Using the decorator pattern seems to be a better idea.

File: cmd/conformance/aws/main.go
Comment (mhutchinson) [reply cmd/conformance/aws/main.go]: Done.

**Pull request merged**
